### PR TITLE
Variable Root width & height

### LINKED ISSUE #552: Allow variable Root width & height
Hello! Love the concepts behind Tidbyt and Pixlet!

I have a usecase for Pixlet involving generating larger images that aren't intended for the Tidbyt. However it seems that Pixlet has the default width and height hardcoded to 64 and 32 px. 

Would there be interest in a PR that added the ability to change the frame and width from Starlark? 

For example:

```python
load("render.star", "render")

def main():
    return render.Root(
        child = render.Text("Hello, World!"),
        frame_width = 128,
        frame_height = 32,
    )

```

Although for some examples there are some changes needed to work with variable dimentions, the examples I tried work fine with variable width and height.

font-preview

Roughly here are the changes if you're interested in a PR:

```diff
diff --git a/render/root.go b/render/root.go
index 0f780ba..3286274 100644
--- a/render/root.go
+++ b/render/root.go
@@ -9,16 +9,14 @@ import (
 	"github.com/tidbyt/gg"
 )
 
-const (
-	// DefaultFrameWidth is the normal width for a frame.
-	DefaultFrameWidth = 64
+// // DefaultFrameWidth is the normal width for a frame.
+var DefaultFrameWidth = 64
 
-	// DefaultFrameHeight is the normal height for a frame.
-	DefaultFrameHeight = 32
+// // DefaultFrameHeight is the normal height for a frame.
+var DefaultFrameHeight = 32
 
-	// DefaultMaxFrameCount is the default maximum number of frames to render.
-	DefaultMaxFrameCount = 2000
-)
+// DefaultMaxFrameCount is the default maximum number of frames to render.
+var DefaultMaxFrameCount = 2000
 
 // Every Widget tree has a Root.
 //
@@ -37,11 +35,12 @@ const (
 // DOC(Child): Widget to render
 // DOC(Delay): Frame delay in milliseconds
 // DOC(MaxAge): Expiration time in seconds
-//
 type Root struct {
-	Child  Widget `starlark:"child,required"`
-	Delay  int32  `starlark:"delay"`
-	MaxAge int32  `starlark:"max_age"`
+	Child       Widget `starlark:"child,required"`
+	Delay       int32  `starlark:"delay"`
+	MaxAge      int32  `starlark:"max_age"`
+	FrameWidth  int32  `starlark:"frame_width"`
+	FrameHeight int32  `starlark:"frame_height"`
 
 	maxParallelFrames int
 	maxFrameCount     int
@@ -80,6 +79,13 @@ func (r Root) Paint(solidBackground bool, opts ...RootPaintOption) []image.Image
 		r.maxFrameCount = DefaultMaxFrameCount
 	}
 
+	if r.FrameWidth != 0 {
+		DefaultFrameWidth = int(r.FrameWidth)
+	}
+	if r.FrameHeight != 0 {
+		DefaultFrameHeight = int(r.FrameHeight)
+	}
+
 	numFrames := r.Child.FrameCount()
 	if numFrames > r.maxFrameCount {
 		numFrames = r.maxFrameCount
diff --git a/runtime/modules/render_runtime/generated.go b/runtime/modules/render_runtime/generated.go
index 2a8187b..6ccea08 100644
--- a/runtime/modules/render_runtime/generated.go
+++ b/runtime/modules/render_runtime/generated.go
@@ -1070,6 +1070,8 @@ func newRoot(
 		child   starlark.Value
 		delay   starlark.Int
 		max_age starlark.Int
+		frame_width starlark.Int
+		frame_height starlark.Int
 	)
 
 	if err := starlark.UnpackArgs(
@@ -1078,6 +1080,8 @@ func newRoot(
 		"child", &child,
 		"delay?", &delay,
 		"max_age?", &max_age,
+		"frame_width?", &frame_width,
+		"frame_height?", &frame_height,
 	); err != nil {
 		return nil, fmt.Errorf("unpacking arguments for Root: %s", err)
 	}
@@ -1108,6 +1112,18 @@ func newRoot(
 		return nil, err
 	}
 
+	if val, err := starlark.AsInt32(frame_width); err == nil {
+		w.FrameWidth = int32(val)
+	} else {
+		return nil, err
+	}
+
+	if val, err := starlark.AsInt32(frame_height); err == nil {
+		w.FrameHeight = int32(val)
+	} else {
+		return nil, err
+	}
+
 	return w, nil
 }
 
@@ -1117,7 +1133,7 @@ func (w *Root) AsRenderRoot() render.Root {
 
 func (w *Root) AttrNames() []string {
 	return []string{
-		"child", "delay", "max_age",
+		"child", "delay", "max_age", "frame_width", "frame_height",
 	}
 }
 
@@ -1136,6 +1152,14 @@ func (w *Root) Attr(name string) (starlark.Value, error) {
 
 		return starlark.MakeInt(int(w.MaxAge)), nil
 
+	case "frame_width":
+
+		return starlark.MakeInt(int(w.FrameWidth)), nil
+
+	case "frame_height":
+
+		return starlark.MakeInt(int(w.FrameHeight)), nil
+
 	default:
 		return nil, nil
 	}
```

### PR DESCRIPTION
**Pull request by ColdHeat**
This PR adds the `frame_width` and `frame_height` arguments for `render.Root` in Pixlet apps. This allows Pixlet to generate variable size images. 

For example:

```python
load("render.star", "render")

def main():
    return render.Root(
        child = render.Text("Hello, World!"),
        frame_width = 128,
        frame_height = 32,
    )
```

If the arguments are not provided the default of 64 and 32 pixels will still be used so there are no changes needed for any apps. 

The `DefaultFrameWidth` and `DefaultFrameHeight` variables are made global so they can be modified but still be accessible from marquee.go where they're used. 

I have tested this on a few examples already and it works fine but I can adjust this based on your feedback.

I realize now that I need to actually codegen the generated.go file and I will figure that out and put another commit up on this. 

Closes #552

### CODE REVIEW

**Comment by ColdHeat** [comment]
It looks like I actually manually wrote the same code that would have been generated 😅.

@betterengineering `make widgets` actually generates a lot of irrelevent documentation changes. I'm not sure if that's due to my environment or something else. 

image

image

Do you want me to include all those changes or would you rather regenerate the docs on your own or something else?

File: render/root.go [hunk 1]
```diff
@@ -9,22 +9,20 @@ import (
 	"github.com/tidbyt/gg"
 )
 
-const (
-	// DefaultFrameWidth is the normal width for a frame.
-	DefaultFrameWidth = 64
+// DefaultFrameWidth is the normal width for a frame.
+var DefaultFrameWidth = 64
```
Comment (betterengineering) [inline render/root.go:13]: Could you leave these as constants?

File: render/root.go [hunk 2]
```diff
@@ -80,6 +81,13 @@ func (r Root) Paint(solidBackground bool, opts ...RootPaintOption) []image.Image
 		r.maxFrameCount = DefaultMaxFrameCount
 	}
 
+	if r.FrameWidth != 0 {
```
Comment (betterengineering) [inline render/root.go:84]: Maybe something like:
```
frameWidth := DefaultFramWidth
if r.FrameWidth != 0 {
    frameWidth = r.FrameWidth
}
```

Sorry if this feels like a nit, I find it really confusing to have a default that mutates 😅

**Review by betterengineering — COMMENTED** [review]
Thanks for the change @ColdHeat! Going to check with the team to ensure I'm not missing any reason we couldn't support this.

File: render/root.go (hunk 1 above)
Comment (ColdHeat) [reply render/root.go:13]: So the issue was that [`marquee.go`](https://github.com/tidbyt/pixlet/blob/main/render/marquee.go#L78-L82) uses these constants.

I can leave these as constants and then create :

``` golang
var FrameWidth = DefaultFrameWidth
var FrameHeight = DefaultFrameHeight
```

which I could then use in marquee.go:

```
	if m.isVertical() {
		cb = m.Child.PaintBounds(image.Rect(0, 0, FrameWidth, m.Height*10), 0)
		cw = cb.Dy()
		size = m.Height
	} else {
		cb = m.Child.PaintBounds(image.Rect(0, 0, m.Width*10, FrameHeight), 0)
		cw = cb.Dx()
		size = m.Width
	}
```

Would that work?

File: render/root.go (hunk 2 above)
Comment (ColdHeat) [reply render/root.go:84]: No worries, golang is not my primarily language so any feedback is great!

**Comment by betterengineering** [comment]
Ok, we kicked the tires a bit here and feel like the _runtime_ should be the one defining the size, and not the _app_. Basically, the source code shouldn't define it's size, and instead `pixlet render` should take a CLI argument to define the size.

The reason being, if Tidbyt supports additional display sizes anytime in our future, we'd like the same apps to require limited changes to scale to different displays. This has benefits outside of our use case too, as you'd be able to use any app in the community repo out of the box with theoretically minimal to no changes on a different display size. 

Would love to hear your thoughts. I know this would require more changes on your part, which is not our intention 😩

File: render/root.go (hunk 1 above)
Comment (betterengineering) [reply render/root.go:13]: Yes, this works! We may want to clean this up later by passing this information into the render tree, but we can follow up in a future change for something like that.

**Comment by ColdHeat** [comment]
That makes sense and I get the sense that it's probably the better decision. I feel like there might be other questions to ask later down the road since the app would render without knowledge of the screen size unless we pass some kind of global information down.

I had this code ready to go so it will take more time to figure out the CLI code and such but I will try to set aside time for it.

**Comment by ColdHeat** [comment]
@betterengineering I couldn't really find a cleaner way to pass the width and height than with this globals strategy. Let me know if you can think of a better way. 

Also let me know if you'd prefer different flags for the CLI arguments.

**Comment by betterengineering** [comment]
This is very cool. Thanks so much for doing this :heart: Here are some renders I've done with a few minor tweaks to the Sunrise Sunset app. To really do this justice, the Tidbyt team or other willing contributor should follow up with:
- Providing $width and $height as config variables passed to `main` so the app can utilize those variables
- Updating pixlet serve to be able to support arbitrary screen sizes

128x128:
sunrise_sunset_128_128

128x32:
sunrise_sunset

**Review by betterengineering — APPROVED** [review]

**Pull request merged**